File: ios/Runner/Features - Main/Models/Content/Content.swift
//
//  BaseClass.swift
//
//  Created by Hiren on 17/10/19
//  Copyright (c) . All rights reserved.
//

import Foundation
import UIKit

class Content: Codable {
    enum CodingKeys: String, CodingKey {
        case track
        case thumbnail
        case objectid
        case contentdetails
        case title
        case ratingtype
        case poster
        case duration
        case seriesid
        case objectowner
        case details
        case seasonnum
        case objectstatus
        case bandorartist
        case trailer
        case releasedate
        case jobid
        case subcategory
        case publishtime
        case contentlanguage
        case genre
        case category
        case tags
        case objecttag
        case productionyear
        case imdbid
        case pgrating
        case albumid
        case subgenre
        case longdescription
        case episode
        case objecttype
        case endtime
        case numviews
        case shortdescription
        // Previous Key
        case seasoncount
        case episodenum
        case episodecount
        case seriesname
        case itemlist
        case seasonid
        case albumname
        case trackcount
        case skip
        case defaulttitle
        case defaultgenre
        case advisory
        case castncrew
        case parentid
        case partnerid
        case playlead
    }
    var track: Int?
    var thumbnail: String?
    var objectid: String?
    var contentdetails: [Contentdetails]?
    var title: String?
    var ratingtype: String?
    var poster: [Poster]?
    var duration: Int?
    var seriesid: String?
    var objectowner: String?
    var details: Details?
    var seasonnum: Int?
    var objectstatus: String?
    var bandorartist: String?
    var trailer: [Trailer]?
    var releasedate: String?
    var jobid: String?
    var subcategory: String?
    var publishtime: String?
    var contentlanguage: [String]?
    var genre: String?
    var category: String?
    var shortDescription: String?
    var tags: [String]?
    var objecttag: [String]?
    var productionyear: Int?
    var imdbid: String?
    var pgrating: String?
    var albumid: String?
    var subgenre: [String]?
    var longdescription: String?
    var episode: String?
    var objecttype: String?
    var endtime: String?
    // Previous keys
    var seasoncount: Int?
    var episodenum: Int?
    var episodecount: Int?
    var seriesname: String?
    var itemlist: [String]?
    var seasonid: String?
    var albumname: String?
    var numviews: String?
    var trackcount: Int?
    var skipSetttings: [SkipSettings]?
    public var landscapeImage: UIImage? // Local
    public var squareImage: UIImage? // Local
    var defaulttitle: String?
    var defaultgenre: String?
    var advisory: String?
    var castncrew: CastnCrewDetails?
    var parentid: String?
    var partnerid: String?
    var playlead: String?
    required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        track = try? container.decodeIfPresent(Int.self, forKey: .track)
        thumbnail = try? container.decodeIfPresent(String.self, forKey: .thumbnail)
        objectid = try? container.decodeIfPresent(String.self, forKey: .objectid)
        contentdetails = try? container.decodeIfPresent([Contentdetails].self, forKey: .contentdetails)
        title = try? container.decodeIfPresent(String.self, forKey: .title)
        ratingtype = try? container.decodeIfPresent(String.self, forKey: .ratingtype)
        poster = try? container.decodeIfPresent([Poster].self, forKey: .poster)
        duration = try? container.decodeIfPresent(Int.self, forKey: .duration)
        seriesid = try? container.decodeIfPresent(String.self, forKey: .seriesid)
        objectowner = try? container.decodeIfPresent(String.self, forKey: .objectowner)
        details = try? container.decodeIfPresent(Details.self, forKey: .details)
        seasonnum = try? container.decodeIfPresent(Int.self, forKey: .seasonnum)
        objectstatus = try? container.decodeIfPresent(String.self, forKey: .objectstatus)
        bandorartist = try? container.decodeIfPresent(String.self, forKey: .bandorartist)
        trailer = try? container.decodeIfPresent([Trailer].self, forKey: .trailer)
        releasedate = try? container.decodeIfPresent(String.self, forKey: .releasedate)
        jobid = try? container.decodeIfPresent(String.self, forKey: .jobid)
        subcategory = try? container.decodeIfPresent(String.self, forKey: .subcategory)
        publishtime = try? container.decodeIfPresent(String.self, forKey: .publishtime)
        contentlanguage = try? container.decodeIfPresent([String].self, forKey: .contentlanguage)
        genre = try? container.decodeIfPresent(String.self, forKey: .genre)
        category = try? container.decodeIfPresent(String.self, forKey: .category)
        shortDescription = try? container.decodeIfPresent(String.self, forKey: .shortdescription)
        tags = try? container.decodeIfPresent([String].self, forKey: .tags)
        objecttag = try? container.decodeIfPresent([String].self, forKey: .objecttag)
        productionyear = try? container.decodeIfPresent(Int.self, forKey: .productionyear)
        imdbid = try? container.decodeIfPresent(String.self, forKey: .imdbid)
        pgrating = try? container.decodeIfPresent(String.self, forKey: .pgrating)
        albumid = try? container.decodeIfPresent(String.self, forKey: .albumid)
        subgenre = try? container.decodeIfPresent([String].self, forKey: .subgenre)
        longdescription = try? container.decodeIfPresent(String.self, forKey: .longdescription)
        episode = try? container.decodeIfPresent(String.self, forKey: .episode)
        objecttype = try? container.decodeIfPresent(String.self, forKey: .objecttype)
        endtime = try? container.decodeIfPresent(String.self, forKey: .endtime)
        seasoncount = try? container.decodeIfPresent(Int.self, forKey: .seasoncount)
        episodenum = try? container.decodeIfPresent(Int.self, forKey: .episodenum)
        episodecount = try? container.decodeIfPresent(Int.self, forKey: .episodecount)
        seriesname = try? container.decodeIfPresent(String.self, forKey: .seriesname)
        seasonid = try? container.decodeIfPresent(String.self, forKey: .seasonid)
        itemlist = try? container.decodeIfPresent([String].self, forKey: .itemlist)
        albumname = try? container.decodeIfPresent(String.self, forKey: .albumname)
        trackcount = try? container.decodeIfPresent(Int.self, forKey: .trackcount)
        skipSetttings = try? container.decodeIfPresent([SkipSettings].self, forKey: .skip)
        defaulttitle = try? container.decodeIfPresent(String.self, forKey: .defaulttitle)
        defaultgenre = try? container.decodeIfPresent(String.self, forKey: .defaultgenre)
        advisory = try? container.decodeIfPresent(String.self, forKey: .advisory)
        castncrew = try? container.decodeIfPresent(CastnCrewDetails.self, forKey: .castncrew)
        parentid = try? container.decodeIfPresent(String.self, forKey: .parentid)
        partnerid = try? container.decodeIfPresent(String.self, forKey: .partnerid)
        playlead = try? container.decodeIfPresent(String.self, forKey: .playlead)
    }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(numviews, forKey: .numviews)
        try container.encode(track, forKey: .track)
        try container.encode(thumbnail, forKey: .thumbnail)
        try container.encode(objectid, forKey: .objectid)
        try container.encode(contentdetails, forKey: .contentdetails)
        try container.encode(title, forKey: .title)
        try container.encode(ratingtype, forKey: .ratingtype)
        try container.encode(poster, forKey: .poster)
        try container.encode(duration, forKey: .duration)
        try container.encode(seriesid, forKey: .seriesid)
        try container.encode(objectowner, forKey: .objectowner)
        try container.encode(details, forKey: .details)
        try container.encode(seasonnum, forKey: .seasonnum)
        try container.encode(objectstatus, forKey: .objectstatus)
        try container.encode(bandorartist, forKey: .bandorartist)
        try container.encode(trailer, forKey: .trailer)
        try container.encode(releasedate, forKey: .releasedate)
        try container.encode(jobid, forKey: .jobid)
        try container.encode(subcategory, forKey: .subcategory)
        try container.encode(publishtime, forKey: .publishtime)
        try container.encode(contentlanguage, forKey: .contentlanguage)
        try container.encode(genre, forKey: .genre)
        try container.encode(category, forKey: .category)
        try container.encode(shortDescription, forKey: .shortdescription)
        try container.encode(tags, forKey: .tags)
        try container.encode(objecttag, forKey: .objecttag)
        try container.encode(productionyear, forKey: .productionyear)
        try container.encode(imdbid, forKey: .imdbid)
        try container.encode(pgrating, forKey: .pgrating)
        try container.encode(albumid, forKey: .albumid)
        try container.encode(subgenre, forKey: .subgenre)
        try container.encode(longdescription, forKey: .longdescription)
        try container.encode(episode, forKey: .episode)
        try container.encode(objecttype, forKey: .objecttype)
        try container.encode(endtime, forKey: .endtime)
        try container.encode(seasoncount, forKey: .seasoncount)
        try container.encode(episodenum, forKey: .episodenum)
        try container.encode(episodecount, forKey: .episodecount)
        try container.encode(seriesname, forKey: .seriesname)
        try container.encode(seasonid, forKey: .seasonid)
        try container.encode(itemlist, forKey: .itemlist)
        try container.encode(albumname, forKey: .albumname)
        try container.encode(trackcount, forKey: .trackcount)
        try container.encode(skipSetttings, forKey: .skip)
        try container.encode(defaulttitle, forKey: .defaulttitle)
        try container.encode(defaultgenre, forKey: .defaultgenre)
        try container.encode(advisory, forKey: .advisory)
        try container.encode(castncrew, forKey: .castncrew)
        try container.encode(parentid, forKey: .parentid)
        try container.encode(partnerid, forKey: .partnerid)
        try container.encode(playlead, forKey: .playlead)
    }
    var displayTitle: String? {
        var title = ""
        if isEpisode {
            if let seasonnum = seasonNumber, !seasonnum.isEmpty {
                title += "\(seasonnum)"
                title += " "
            }
            if let episodenum = episodeNumber, !episodenum.isEmpty {
                title += "\(episodenum)"
            }
        }
        if title != "" {
            title += ": "
        }
        return title + (self.title ?? "")
    }
}
// MARK: - Series Related
extension Content {
    var isEpisode: Bool {
        return objecttype == "CONTENT" && seriesid != nil && episodenum != nil
    }
    var isSeason: Bool {
        return objecttype == "SEASON" && seriesid != nil
    }
    var seasonNumber: String? {
        var title = ""
        if let seasonnum = seasonnum {
            title += "S" + "\(seasonnum)"
        }
        return title
    }
    private var episodeNumber: String? {
        var title = ""
        if let episodenum = episodenum {
            title += "E" + "\(episodenum)"
        }
        return title
    }
    var episodeTitle: String {
        return "\(seriesname ?? "")" + "\n" + "\(seasonNumber ?? "")" + ": " + "\(title ?? "")"
    }
}
// MARK: - Poster Related
extension Content {
    func fetchImageURL(_ displayType: DisplayType?, imageSize: CGSize) -> String? {
        guard let posterArray = poster, let displayType = displayType else { return thumbnail }
        if displayType == .MIXED {
            return fetchImageURL(.LANDSCAPE, imageSize: imageSize) ?? fetchImageURL(.PORTRAIT, imageSize: imageSize) ?? fetchImageURL(.SQUARE, imageSize: imageSize) ?? thumbnail
        } else {
            if let selectedPoster = posterArray.filter({ (poster) -> Bool in
                return poster.postertype == displayType.rawValue
            }).first {
                let sdImage = selectedPoster.filelist?.filter({ (file) -> Bool in
                    return file.quality == "SD"
                }).first?.filename
                let lowImage = selectedPoster.filelist?.filter({ (file) -> Bool in
                    return file.quality == "LOW"
                }).first?.filename
                let thumbnailImage = selectedPoster.filelist?.filter({ (file) -> Bool in
                    return file.quality == "THUMBNAIL"
                }).first?.filename
                let hdImage = selectedPoster.filelist?.filter({ (file) -> Bool in
                    return file.quality == "HD"
                }).first?.filename
                if imageSize == .zero {
                    return sdImage ?? lowImage ?? thumbnailImage ?? hdImage
                } else if displayType == .LANDSCAPE {
                    if imageSize.width <= 256 {
                        return thumbnailImage ?? lowImage ?? sdImage
                    } else if imageSize.width <= 480 {
                        return lowImage ?? sdImage
                    } else if imageSize.width <= 960 {
                        return sdImage
                    } else {
                        return hdImage
                    }
                } else if displayType == .PORTRAIT {
                    if imageSize.width <= 170 {
                        return thumbnailImage ?? lowImage ?? sdImage
                    } else if imageSize.width <= 320 {
                        return lowImage ?? sdImage
                    } else if imageSize.width <= 640 {
                        return sdImage
                    } else {
                        return hdImage
                    }
                } else if displayType == .SQUARE {
                    if imageSize.width <= 256 {
                        return thumbnailImage ?? lowImage ?? sdImage ?? hdImage
                    } else if imageSize.width <= 480 {
                        return lowImage ?? sdImage ?? hdImage
                    } else if imageSize.width <= 960 {
                        return sdImage ?? hdImage
                    } else {
                        return hdImage
                    }
                }
            }
            return thumbnail
        }
    }
    
}
// MARK: - Skip Settings
extension Content {
    func fetchSkipIntro() -> (Double, Double)? {
        if let skipIntro = skipSetttings?.filter({ (settings) -> Bool in
            return settings.skiptype == "INTRO"
        }).first, let start = skipIntro.start, let end = skipIntro.end {
            return (start, end)
        }
        return nil
    }
    func fetchNextEpisode() -> (Double, Double)? {
        if let skipIntro = skipSetttings?.filter({ (settings) -> Bool in
            return settings.skiptype == "CREDITS"
        }).first, let start = skipIntro.start, let end = skipIntro.end {
            return (start, 10) // end hardcoded as for now
        }
        return nil
    }
    var isClearLeadEnabled: Bool {
        return playlead == "YES"
    }
}

File: android/app/src/main/kotlin/com/enrichtv/android/player/models/content/Content.kt
/**
 * Model class for Content data
 * */
@file:Suppress("SpellCheckingInspection")

package com.enrichtv.android.player.models.content

import android.os.Parcelable
import androidx.annotation.Keep
import com.enrichtv.android.player.ApiConstant.ALBUM_ID
import com.enrichtv.android.player.ApiConstant.BANDOR_ARTIST
import com.enrichtv.android.player.ApiConstant.CATEGORY
import com.enrichtv.android.player.ApiConstant.CITY
import com.enrichtv.android.player.ApiConstant.CONTENT
import com.enrichtv.android.player.ApiConstant.CONTENT_LANGUAGE
import com.enrichtv.android.player.ApiConstant.CONTEN_TOWNER
import com.enrichtv.android.player.ApiConstant.DETAILS
import com.enrichtv.android.player.ApiConstant.DURATION
import com.enrichtv.android.player.ApiConstant.EPISODE_COUNT
import com.enrichtv.android.player.ApiConstant.EPISODE_NUM
import com.enrichtv.android.player.ApiConstant.GENRE
import com.enrichtv.android.player.ApiConstant.HD
import com.enrichtv.android.player.ApiConstant.IS_BLANK
import com.enrichtv.android.player.ApiConstant.ITEM_LIST
import com.enrichtv.android.player.ApiConstant.LOW
import com.enrichtv.android.player.ApiConstant.NUM_VIEWS
import com.enrichtv.android.player.ApiConstant.OBJECT_ID
import com.enrichtv.android.player.ApiConstant.OBJECT_TAG
import com.enrichtv.android.player.ApiConstant.OBJECT_TYPE
import com.enrichtv.android.player.ApiConstant.PG_RATING
import com.enrichtv.android.player.ApiConstant.POSTER
import com.enrichtv.android.player.ApiConstant.PRODUCTION_YEAR
import com.enrichtv.android.player.ApiConstant.RATING
import com.enrichtv.android.player.ApiConstant.RATING_TYPE
import com.enrichtv.android.player.ApiConstant.RELEASE_DATE
import com.enrichtv.android.player.ApiConstant.SD
import com.enrichtv.android.player.ApiConstant.SEASON_COUNT
import com.enrichtv.android.player.ApiConstant.SEASON_NUM
import com.enrichtv.android.player.ApiConstant.SERIES_ID
import com.enrichtv.android.player.ApiConstant.SERIES_NAME
import com.enrichtv.android.player.ApiConstant.SHORT_DESCRIPTION
import com.enrichtv.android.player.ApiConstant.SUBSCRIBER_ID
import com.enrichtv.android.player.ApiConstant.SUB_CATEGORY
import com.enrichtv.android.player.ApiConstant.SUB_GENRE
import com.enrichtv.android.player.ApiConstant.TAGS
import com.enrichtv.android.player.ApiConstant.THUMBNAIL
import com.enrichtv.android.player.ApiConstant.TILL_DATE
import com.enrichtv.android.player.ApiConstant.TITLE
import com.enrichtv.android.player.ApiConstant.TRACK
import com.enrichtv.android.player.ApiConstant.TRACK_COUNT
import com.enrichtv.android.player.ApiConstant.TRAILER
import com.enrichtv.android.player.ApiConstant.VENUE

import com.google.gson.annotations.SerializedName
import com.mobiotics.core.extensions.isNotNull
import com.mobiotics.core.extensions.isNull
import com.enrichtv.android.player.ApiConstant.LONG_DESCRIPTION
import com.enrichtv.android.player.ApiConstant.POSTER_LINK
import com.enrichtv.android.player.ApiConstant.SKIP
import com.enrichtv.android.player.Constants.EMPTY_STRING
import com.enrichtv.android.player.Constants.MINUS_DEFAULT
import com.enrichtv.android.models.Stream
import com.enrichtv.android.models.config.SubContents
import com.enrichtv.android.models.content.Skip
import kotlinx.android.parcel.Parcelize

@Parcelize
@Keep
data class Content(
    @SerializedName(OBJECT_ID)
    var objectid: String = "",
    @SerializedName(ALBUM_ID)
    var albumid: String? = null,
    @SerializedName(BANDOR_ARTIST)
    var bandOrArtist: String? = null,
    @SerializedName(CATEGORY)
    var category: CategoryType? = null,
    @SerializedName(CONTENT, alternate = ["contentdetails"])
    var listDetails: List<ContentDetails>? = listOf(),
    @SerializedName(CONTENT_LANGUAGE)
    var contentLanguage: List<String>? = null,
    @SerializedName(CONTEN_TOWNER)
    var contentOwner: String? = null,
    @SerializedName(DETAILS)
    var details: Details? = null,
    @SerializedName(DURATION)
    var duration: Int = 0,
    @SerializedName(EPISODE_NUM)
    var episodeNum: String? = null,
    @SerializedName(GENRE)
    var genre: String? = null,
    @SerializedName(EPISODE_COUNT)
    var episodeCount: Int? = null,
    @SerializedName(NUM_VIEWS)
    var numviews: String? = null,
    @SerializedName(OBJECT_TYPE)
    var objectType: ObjectType? = null,
    @SerializedName(PG_RATING)
    var pgRating: String? = null,
    @SerializedName(POSTER)
    var poster: List<Poster>? = listOf(),
    @SerializedName(PRODUCTION_YEAR)
    var productionYear: String? = null,
    @SerializedName(RATING)
    var rating: Rating? = null,
    @SerializedName(RATING_TYPE)
    var ratingType: String? = null,
    @SerializedName(RELEASE_DATE)
    var releaseDate: String? = null,
    @SerializedName(SEASON_NUM)
    var seasonNum: Int? = null,
    @SerializedName(SEASON_COUNT)
    var seasonCount: String? = null,
    @SerializedName(SERIES_ID)
    var seriesid: String? = null,
    @SerializedName(SHORT_DESCRIPTION)
    var shortDescription: String? = null,
    @SerializedName(SUB_GENRE)
    var subgenre: List<String>? = null,
    @SerializedName(TAGS)
    var tags: List<String>? = null,
    @SerializedName(TILL_DATE)
    var tillDate: String? = null,
    @SerializedName(TITLE)
    var title: String? = null,
    @SerializedName(TRACK)
    var track: String? = null,
    @SerializedName(SUBSCRIBER_ID)
    var subscriberId: String? = null,
    @SerializedName(TRAILER)
    var trailer: List<Trailer>?,
    @SerializedName(IS_BLANK)
    var isBlank: Boolean = false,
    @SerializedName(ITEM_LIST)
    var itemList: List<String>? = listOf(),
    @SerializedName(SERIES_NAME)
    var seriesName: String? = null,
    var isCurrentseries: Boolean = false,
    var profileId: String? = null,
    var videoSize: Long? = null,
    @SerializedName(THUMBNAIL)
    val thumbnail: String? = null,
    @SerializedName(OBJECT_TAG)
    var objectTag: List<String>? = null,
    @SerializedName(SUB_CATEGORY)
    val subCategory: String? = null,
    @SerializedName(TRACK_COUNT)
    val trackcount: String? = null,
    @SerializedName(VENUE)
    val venue: String? = null,
    @SerializedName(CITY)
    val city: String? = null,
    @SerializedName(LONG_DESCRIPTION)
    val longDescription: String? = null,
    @SerializedName(SKIP)
    val skip: List<Skip>? = null,
    @SerializedName("watchedDuration")
    var watchedDuration: Long? = null,
    @SerializedName("playlead")
    var playlead: String? = null,

    var description: String? = null,
    var objectstatus: ObjectStatus? = null,
    var jobid: String? = null,
    var imdbid: String? = null,
    @SerializedName("contentStream")
    var contentStream: Stream? = null,
    var isTrailer: Boolean = false,
    var seasonid: String? = null,
    var subContents: SubContents? = null,
    @SerializedName(POSTER_LINK)
    var posterLink: String? = null,
    var isAdded: Boolean = false,
    var defaulttitle: String? = null,
    var defaultgenre: String? = null,
    var advisory: String? = null,
    var metacontent: MetaContent? = null,
    var castncrew: CastCrew? = null,
    var isChecked: Boolean = false,


    ) : Parcelable {

    /**
     * When pass a blank object use this constructor
     *
     * @modifier Sudhir Kumar
     * @since 31/03/2020
     */
    @Keep
    constructor(isBlank: Boolean) : this(
        EMPTY_STRING,
        EMPTY_STRING,
        EMPTY_STRING,
        null,
        emptyList(),
        null,
        EMPTY_STRING,
        null,
        MINUS_DEFAULT,
        null,
        EMPTY_STRING,
        null,
        EMPTY_STRING,
        null,
        EMPTY_STRING,
        emptyList<Poster>(),
        EMPTY_STRING,
        Rating(MINUS_DEFAULT),
        EMPTY_STRING,
        null,
        0,
        EMPTY_STRING,
        EMPTY_STRING,
        EMPTY_STRING,
        null,
        emptyList<String>(),
        EMPTY_STRING,
        EMPTY_STRING,
        EMPTY_STRING,
        EMPTY_STRING,
        null,
        isBlank,
        null,
        null,
        false,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        skip = null,
        advisory = null,
        metacontent = null,
        castncrew = null
    )
}

@Keep
enum class Orientation {
    PORTRAIT, LANDSCAPE, SQUARE
}


fun Content.getTrailer(): Trailer? {
    trailer?.forEach { trailer ->
        for (quality in SORT_ORDER) {
            if (trailer.quality == quality) {
                return trailer
            }
        }
    }
    return null
}


fun Trailer.getFileList(): FileList? {
    filelist?.forEach { fileList ->
        for (quality in SORT_ORDER) {
            if (fileList.quality == quality) {
                return fileList
            }
        }
    }
    return null
}

internal var SORT_ORDER = arrayOf(SD, LOW, THUMBNAIL, HD)

fun Content.poster(orientation: Orientation): String? {
    var posterData = poster?.find { it.postertype == orientation }
    if (posterData.isNull()) {
        posterData =
            poster?.find { it.postertype == Orientation.PORTRAIT || it.postertype == Orientation.LANDSCAPE || it.postertype == Orientation.SQUARE }
    }
    val fileList = posterData?.fileList
    if (fileList.isNullOrEmpty()) return null
    val sorteList = if (fileList.size == 1) fileList
    else {
        val mList = fileList.toMutableList()
        val index = mList.indexOfFirst { it.quality == HD }
        if (index != -1) {
            val hd = mList.removeAt(index)
            mList.add(hd)
        }
        mList
    }
    return sorteList.firstOrNull()?.fileName
    /*val list: List<Poster>? = poster?.filter { it.postertype == orientation }
    list?.forEach { poster ->
        for (quality in SORT_ORDER) {
            val fileList =
                poster.fileList?.find { posterQuality -> posterQuality.quality == quality }
            if (fileList != null)
                return fileList.fileName
        }
    }
    return list?.find { it ->
        it.fileList?.find { it.quality.equals(Constants.QUALITY_HD, true) }.isNotNull()
    }?.fileList?.first()?.fileName*/
}

fun createAdditionalParamForDetailApi(
    country: String? = null,
    region: String? = null,
    description: String? = null
): MutableMap<String, String?> {
    val mapParam: MutableMap<String, String?> = mutableMapOf()
    mapParam.apply {
        if (country.isNotNull()) put("country", country)
        if (region.isNotNull()) put("region", region)
        if (description.isNotNull()) put("description", description)
    }
    return mapParam
}

/**
 * Returns whether the Content type is Music or not.
 *
 * @return true if music; false otherwise
 */
fun Content?.isTypeMusic(): Boolean {
    if (this == null) return false
    return objectType == ObjectType.ALBUM && category == CategoryType.MUSIC
}

/**
 * Returns whether the Content type is Online Radio or not.
 *
 * @return true if online radio; false otherwise.
 */
fun Content?.isTypeRadio(): Boolean {
    if (this == null) return false
    return (objectType == ObjectType.CHANEL || category == CategoryType.ONLINE) && genre == "RADIO"
}
